### PR TITLE
Adaptor updates - #2500 and maybe more

### PR DESCRIPTION
**Pull request by evs-chris**
**Description of the pull request:**
This:

* consolidates all of the adaptor handling in model into the `adapt` method
* adds a check in `ractive.update` for an adapted parent and delegates the update to the parent if it is adapted
* adds reset methods to all of the provided adaptors that don't cause them to teardown if set with their current value

I think that covers the rest of the weird cases of adaptors, but I'm not sure as I haven't used them extensively. This may also fix some of the magic failures in edge, as it adds a working `reset`. At least I think it is a working `reset`... can anyone comment on the implementation?

Have I missed or completely misunderstood anything?

**Fixes the following issues:**
#2500 

**Is breaking:**
I don't think so

**Reviewers:**
Anyone familiar with how adaptors should behave

### CODE REVIEW

File: src/Ractive/static/adaptors/array/index.js [hunk 1]
```diff
@@ -50,6 +50,11 @@ class ArrayWrapper {
 		return this.value;
 	}
 
+	reset ( value ) {
+		if ( this.value === value ) return true;
+		return false;
+	}
```
Comment (fskreuz) [inline src/Ractive/static/adaptors/array/index.js]:
```js
reset ( value ){
  return this.value === value;
}
```

Comment (evs-chris) [inline src/Ractive/static/adaptors/array/index.js]: :+1: I had a stashing derp whilst testing something just before commit.

**Comment by Jeff17Robbins** [comment]
We have been using https://github.com/rstacruz/ractive-ractive in Ractive.js v0.7.3, and would like to do so in Ractive.js v0.8.0.  

Aside from the update issues I've been trying to reproduce (and @evs-chris  has been fixing -- thank you!), there's also issue #2090.  

The Ractive "change" event doesn't fire for template-less Ractive instances.  (In fact, it appears that ractive.viewmodel.changes accumulates the changes for a template-less ractive instance, but `flushChanges` doesn't handle the template-less or lacking-a-fragment case, and so the viewmodel.changes simply accumulate, never to be fired via `changeHook.fire(ractive, ractive.viewmodel.changes);`.)

So...if the https://github.com/rstacruz/ractive-ractive adaptor is intended to work in Ractive.js v0.8.0, the Ractive lifecycle event `change` needs to be fixed to work on template-less Ractive instances.  

This comment likely doesn't really belong on this PR, but I don't know how to flag the whole adaptor topic, and would like to help with making adaptors work.

**Comment by aharpervc** [comment]
I agree with @Jeff17Robbins that ractive-ractive is a huge boon and would be a giant problem if there was no way to achieve a similar result in 0.8.0 (which might as well be called v1 at this point :skull: )

**Comment by martypdx** [comment]
@evs-chris I think there are three things that adaptors get used for, it might be worth separating them out:

1. Mediating between a data source and the ractive instance model. Most classic use and what people generally think of as adaptors. Your spot on that we need to have dedicated array modification methods. One of the other needed additions to delineate requests made for child properties for the purpose of access a further child property versus for a value to be used in the template. For me, this came up with working with Firebase as getting a childen reference has a much different effect than getting a child value (the former is a reference, the later fetches data from the server).
2. Mediating between ractive model and the template. For the most part, this should be handled by computed properties, but I've seen it crop up as an adaptor when devs try to apply to a testable conditions, for example show any boolean value as Y or N. If it's between the model and template, IMO it shouldn't be an adaptor
3. Extending the ractive API. The data needs a way to tell ractive about changes, the inverse of #1. Usually this would be in conjunction with first point, but interestingly both the array and magic adaptors fall in this category, but not the first.

**Comment by Jeff17Robbins** [comment]
@martypdx thank you for the three use cases.  I thought it might be worth sharing what we have been doing with the [Ractive-Ractive adaptor](https://github.com/rstacruz/ractive-ractive).  

We have been using it to create a template-less Ractive instance that acts as a "store" to several other Ractive instances that act as "views".  We use it one-way only, meaning that we only need data flowing one-way from the store to the views.

While there is nothing mandating that we use a Ractive instance as our store, it seems handy because we then also get computeds in our store, for free, so to speak.  Since many views can share one store, it seems clean to us.  

If there is another way to manage a single external model, and somehow map it into models of our views, we could use that other way. The basic requirements are unsurprising:

1. A single location to deposit incoming JSON data.  
2. A way to then update **efficiently** several different Ractive instances with that data.

Our views use mustaches like `{{store.somevalue}}` or `{{store.someobj.someprop}}` in their templates.  The initial part of the path 'store' is meant to pick out the store Ractive instance.

In some ways, our use is like your use-case 1, except we want a staging area (aka "store") to first deposit incoming data so that we can then have many Ractive instances ("views") sharing this data.  But in another sense, it seems like your use-case 3, in that when our store gets new data, we need it to tell ractive (the ractive instances acting as "views") about the changes.  If we can continue to use the Ractive-Ractive adaptor, that would be easiest for us.  But if there is a better (or equivalent) way to model this in [Ractive.js v0.8.0](http://cdn.ractivejs.org/edge/ractive.js), we are totally open to a different approach.

Any suggestions?

**Comment by evs-chris** [comment]
I think I've managed to mangle the various adaptor scenarios into working, but it could certainly use some review. Adaptors are now a bit more expensive during shuffles than they used to be because the model now tears adaptors on shuffled paths down so that they don't get out of whack with further updates. So the test for #2005 is now passing in this PR too.

This solution is probably not ideal, but I'm pretty sure a more reasonable solution involves breaking the adaptor API.

**Comment by sabob** [comment]
Looks good.

**Pull request merged**